central/blob/datastore/store/store.go

```diff
@@ -0,0 +1,177 @@
+package store
+
+import (
+	"context"
+	"io"
+
+	"github.com/jackc/pgx/v4"
+	"github.com/pkg/errors"
+	"github.com/stackrox/rox/central/blob/datastore/store/postgres"
+	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/logging"
+	pgPkg "github.com/stackrox/rox/pkg/postgres"
+)
+
+var log = logging.LoggerForModule()
+
+// Store is the interface to interact with the storage for storage.Blob
+type Store interface {
+	Upsert(ctx context.Context, obj *storage.Blob, reader io.Reader) error
+	Get(ctx context.Context, name string, writer io.Writer) (*storage.Blob, bool, error)
+	Delete(ctx context.Context, name string) error
+}
+
+type storeImpl struct {
+	db    pgPkg.DB
+	store postgres.Store
+}
+
+// New creates a new Blob store
+func New(db pgPkg.DB) Store {
+	return &storeImpl{
+		db:    db,
+		store: postgres.New(db),
+	}
+}
+
+func wrapRollback(ctx context.Context, tx *pgPkg.Tx, err error) error {
+	rollbackErr := tx.Rollback(ctx)
+	if rollbackErr != nil {
+		return errors.Wrapf(rollbackErr, "rolling back due to err: %v", err)
+	}
+	return err
+}
+
+// Upsert adds a blob to the database
+func (s *storeImpl) Upsert(ctx context.Context, obj *storage.Blob, reader io.Reader) error {
+	existingBlob, exists, err := s.store.Get(ctx, obj.GetName())
+	if err != nil {
+		return err
+	}
+	tx, err := s.db.Begin(ctx)
+	if err != nil {
+		return err
+	}
+	ctx = pgPkg.ContextWithTx(ctx, tx)
+
+	los := tx.LargeObjects()
+	var lo *pgx.LargeObject
+	if exists {
+		lo, err = los.Open(ctx, existingBlob.GetOid(), pgx.LargeObjectModeWrite)
+		if err != nil {
+			return wrapRollback(ctx, tx, errors.Wrapf(err, "opening blob with oid %d", existingBlob.GetOid()))
+		}
+		if err := lo.Truncate(0); err != nil {
+			return errors.Wrapf(err, "truncating blob with oid %d", existingBlob.GetOid())
```

Comment:
Since we are neither rolling back or committing, won't this leak the transaction?